block-scoped-var shows errors for object properties

Lint this file wit `"block-scoped-var": 2`

``` javascript
function generateDefinition(config) {
    var definition = {};
    if (config.axisType) {
        definition.x = { types: [config.axisType], filter: config.axisFilter };
    }
    if (config.cellType) {
        definition.cells = { types: [].concat(config.cellType) };
    }
    return definition;
}
```

Result:

```
  3:15  error  "axisType" used outside of binding context    block-scoped-var
  4:19  error  "x" used outside of binding context           block-scoped-var
  4:25  error  "types" used outside of binding context       block-scoped-var
  4:40  error  "axisType" used outside of binding context    block-scoped-var
  4:51  error  "filter" used outside of binding context      block-scoped-var
  4:66  error  "axisFilter" used outside of binding context  block-scoped-var
  6:15  error  "cellType" used outside of binding context    block-scoped-var
  7:19  error  "cells" used outside of binding context       block-scoped-var
  7:29  error  "types" used outside of binding context       block-scoped-var
  7:39  error  "concat" used outside of binding context      block-scoped-var
  7:53  error  "cellType" used outside of binding context    block-scoped-var
```

Expected result:
no errors

--- a/lib/rules/block-scoped-var.js
+++ b/lib/rules/block-scoped-var.js
@@ -23,16 +23,30 @@ module.exports = function(context) {
      * @returns {Boolean} true when the identifier is in declaration position.
      */
     function isDeclaration(id, parent) {
-        switch(parent.type) {
+        switch (parent.type) {
             case "FunctionDeclaration":
             case "FunctionExpression":
                 return parent.params.indexOf(id) > -1 || id === parent.id;
+
             case "VariableDeclarator":
                 return id === parent.id;
+
             case "CatchClause":
                 return id === parent.param;
+
+            default:
+                return false;
         }
-        return false;
+    }
+
+    /**
+     * Determines whether an identifier is in property position.
+     * @param {ASTNode} id The identifier.
+     * @param {ASTNode} parent The identifier's parent AST node.
+     * @returns {Boolean} true when the identifier is in property position.
+     */
+    function isProperty(id, parent) {
+        return id === parent.property && parent.type === "MemberExpression" && !parent.computed;
     }
 
     /**
@@ -66,13 +80,17 @@ module.exports = function(context) {
 
     function functionHandler(node) {
         pushScope();
-        declare(node.params.map(function(id) { return id.name; }));
+        declare(node.params.map(function(id) {
+            return id.name;
+        }));
         declare(node.id ? [node.id.name] : []);
     }
 
     return {
         "Program": function() {
-            scopeStack = [context.getScope().variables.map(function(v) { return v.name; })];
+            scopeStack = [context.getScope().variables.map(function(v) {
+                return v.name;
+            })];
         },
 
         "BlockStatement": function(node) {
@@ -80,12 +98,15 @@ module.exports = function(context) {
             pushScope();
             statements.forEach(function(stmt) {
                 if (stmt.type === "VariableDeclaration") {
-                    declare(stmt.declarations.map(function(decl) { return decl.id.name; }));
+                    declare(stmt.declarations.map(function(decl) {
+                        return decl.id.name;
+                    }));
                 } else if (stmt.type === "FunctionDeclaration") {
                     declare([stmt.id.name]);
                 }
             });
         },
+
         "BlockStatement:exit": popScope,
 
         "CatchClause": function(node) {
@@ -97,7 +118,10 @@ module.exports = function(context) {
         "FunctionExpression": functionHandler,
 
         "Identifier": function(node) {
-            if (isDeclaration(node, context.getAncestors().pop())) { return; }
+            var ancestor = context.getAncestors().pop();
+            if (isDeclaration(node, ancestor) || isProperty(node, ancestor)) {
+                return;
+            }
             for (var i = 0, l = scopeStack.length; i < l; i++) {
                 if (scopeStack[i].indexOf(node.name) > -1) {
                     return;

--- a/tests/lib/rules/block-scoped-var.js
+++ b/tests/lib/rules/block-scoped-var.js
@@ -29,7 +29,8 @@ eslintTester.addRuleTest("lib/rules/block-scoped-var", {
         "function f(){ g(); function g(){} }",
         { code: "new Date", globals: {Date: false} },
         { code: "new Date", globals: {} },
-        { code: "var eslint = require('eslint');", globals: {require: false} }
+        { code: "var eslint = require('eslint');", globals: {require: false} },
+        "function f(a) { return a.b; }"
     ],
     invalid: [
         { code: "function f(){ x; }", errors: [{ message: "\"x\" used outside of binding context.", type: "Identifier" }] },
@@ -37,6 +38,8 @@ eslintTester.addRuleTest("lib/rules/block-scoped-var", {
         { code: "function f(){ { var x; } x; }", errors: [{ message: "\"x\" used outside of binding context.", type: "Identifier" }] },
         { code: "function f() { var a; { var b = 0; } a = b; }", errors: [{ message: "\"b\" used outside of binding context.", type: "Identifier" }] },
         { code: "function f() { try { var a = 0; } catch (e) { var b = a; } }", errors: [{ message: "\"a\" used outside of binding context.", type: "Identifier" }] },
-        { code: "var eslint = require('eslint');", globals: {}, errors: [{ message: "\"require\" used outside of binding context.", type: "Identifier" }] }
+        { code: "var eslint = require('eslint');", globals: {}, errors: [{ message: "\"require\" used outside of binding context.", type: "Identifier" }] },
+        { code: "function f(a) { return a[b]; }", errors: [{ message: "\"b\" used outside of binding context.", type: "Identifier" }] },
+        { code: "function f() { return b.a; }", errors: [{ message: "\"b\" used outside of binding context.", type: "Identifier" }] }
     ]
 });

